src/sentry/release_health/tasks.py

```diff
@@ -179,6 +164,69 @@ class AdoptedRelease(TypedDict):
     version: str
 
 
+def query_adopted_release_project_environments(
+    adopted_releases: list[AdoptedRelease],
+) -> list[ReleaseProjectEnvironment]:
+    """Fetch a list of release project environment rows which match the adopted releases."""
+    query_filters = Q()
+    for release in adopted_releases:
+        query_filters |= Q(
+            release__version=release["version"],
+            project_id=release["project_id"],
+            environment__name=release["environment"],
+        )
+
+    return list(ReleaseProjectEnvironment.objects.filter(query_filters))
+
+
+def adopt_release_project_environments(models: list[ReleaseProjectEnvironment]) -> None:
+    """Bulk update release project environments with adopted time."""
+    for model in models:
+        if not model.adopted:
+            model.adopted = timezone.now()
+        if model.unadopted is not None:
+            model.unadopted = None
+
+    ReleaseProjectEnvironment.objects.bulk_update(models, fields=["adopted", "unadopted"])
+
+
+def find_adopted_but_missing_releases(
+    adopted_releases: list[AdoptedRelease],
+    found_rows: list[tuple[int, str, str]],
+) -> list[AdoptedRelease]:
+    """Return a list of adopted releases which do not exist in the database.
+
+    Releases are not always defined in Sentry in advance. When we see a release that isn't in the
+    database we need to do some additional processing.
+    """
+
+    def hash_release(q: AdoptedRelease) -> tuple[int, str, str]:
+        return (q["project_id"], q["environment"], q["version"])
+
+    return find_missing(adopted_releases, hash_release, found_rows, lambda a: a)
+
+
+A = TypeVar("A")
+B = TypeVar("B")
+C = TypeVar("C")
+D = TypeVar("D")
+
+
+def find_missing(
+    superset: list[A],
+    superset_hasher: Callable[[A], B],
+    subset: list[C],
+    subset_hasher: Callable[[C], D],
+) -> list[A]:
```

Comment:
shouldn't `B` = `D` otherwise this isn't typesafe?

Reply:
@asottile-sentry Ran mypy with `--strict-equality` and it detected it.  Forgot that mypy had strictness levels 😓